### ISSUE TITLE
Client hangs forever if amount of retries >= maxTotalConnections

### ISSUE THREAD

**Issue by mauriciojost** [http4s/http4s#2677]
I am using http4s 0.20.3.
My client uses http4s to perform a PUT on an external server. The external server (not in scope) replies with 503 Service Unavailable under certain conditions. If the amount of queries failed with such code == maxTotalConnections in BlazeClientBuilder then my application hangs.

It seems very similar to http4s/http4s#2068, but instead of sending requests in parallel I need N failed queries for the application to hang.

Furthermore, if I have maxTotalConnections=20 and I request 23 retries, the application will retry 20 times and hang.

**Comment by nigredo-tori** [comment]
That's because the implementation [`.flatMap`s `Resource`s](https://github.com/http4s/http4s/blob/4b92d3712c55dfa8c04219214159c8c9474fd4f1/client/src/main/scala/org/http4s/client/middleware/Retry.scala#L42) on failure. So if we do 20 tries, we do 20 acquires before the first release can happen.

The necessary behavior looks difficult to implement with `Resource` itself (unless we drop down to `.allocated` - and that's dangerous). It might be easier to solve this with a properly `.scope`d `fs2.Stream`.

**Comment by nigredo-tori** [comment]
Hm, it seems that with `.compile.resource.last` `Stream` releases `.scope`d resources before yielding. No dice.

**Comment by rossabaker** [comment]
In 0.18, before we used `Resource`, we would [`dispose`](https://github.com/http4s/http4s/blob/v0.18.24/client/src/main/scala/org/http4s/client/middleware/Retry.scala#L44) before a next attempt.

I think you may be right about `allocated`, but that's going to be a bit scary with cancellation semantics.

**Comment by rossabaker** [comment]
It's also an issue in `FollowRedirect`.

**Comment by rossabaker** [comment]
Here's a unit test for it.  (I didn't appreciate what a pain it was to create this middleware.)

```scala
    "not exhaust the connection pool on retry" in {
      Semaphore[IO](2).flatMap { semaphore =>
        val client = Retry[IO](RetryPolicy((att =>
          if (att < 3) Some(Duration.Zero)
          else None
        ), RetryPolicy.defaultRetriable[IO]))(Client[IO](_ =>
          Resource.make(semaphore.tryAcquire.flatMap {
            case true => Response[IO](Status.InternalServerError).pure[IO]
            case false => IO.raiseError(new IllegalStateException("Exhausted all connections"))
          })(_ => semaphore.release)
        ))
        client.status(Request[IO]())
      } must returnValue(Status.InternalServerError)
    }
```
